File: skip_list.go
// resource : https://mecha-mind.medium.com/redis-sorted-sets-and-skip-lists-4f849d188a33
package main

import (
	"fmt"
	"math"
	"math/rand"
	"time"
)

type Sample struct {
	values  *[]int
	weights *[]float64
	cumSum  *[]float64
}

func (samp *Sample) init(values *[]int, weights *[]float64) {
	sumWeights := (func(arr *[]float64) float64 {
		tempSum := 0.0
		for _, v := range *arr {
			tempSum += v
		}
		return tempSum
	})(weights)
	tempCumSum := make([]float64, len(*values))

	samp.values = values
	samp.weights = (func(weights *[]float64, weightSum float64) *[]float64 {
		tempWeights := make([]float64, len(*weights))
		if weightSum != 0 {
			for index, elem := range *weights {
				tempWeights[index] = elem / weightSum
			}
		}
		return &tempWeights
	})(weights, sumWeights)
	samp.cumSum = &tempCumSum
	samp.computerCumSum()
}
func (samp *Sample) computerCumSum() {
	s := 0.0
	for index, w := range *samp.weights {
		s += w
		(*samp.cumSum)[index] += s
	}
}
func (samp *Sample) getOne() int {
	src := rand.NewSource(time.Now().UnixNano())
	rng := rand.New(src)
	randProbability := rng.Float64()

	// perform binary search lol
	left, right, lastTrue := 0, len(*samp.values)-1, 0

	for left <= right {
		mid := left + ((right - left) / 2)

		if (*samp.cumSum)[mid] >= randProbability {
			lastTrue = mid
			right = mid - 1
		} else {
			left = mid + 1
		}
	}
	return (*samp.values)[lastTrue]
}

type SLNode struct {
	key          *string // int or nil
	val          *int    // string or nil
	nextPointers map[int]*SLNode
	prevPointers map[int]*SLNode
	level        *int // int or nil
}

func (sln *SLNode) init(key *string, val *int, level *int) {
	sln.key = key
	sln.val = val
	sln.level = level
	sln.nextPointers = make(map[int]*SLNode)
	sln.prevPointers = make(map[int]*SLNode)
}

type SkipList struct {
	levels        *[]int
	probabilities *[]float64
	sample        *Sample
	head          *SLNode
	tail          *SLNode
	nodeMap       map[string]*SLNode
}

func (sl *SkipList) init(maxSize int) {

	tempLevels := make([]int, 0)
	for i := 1; i < int(math.Log2(float64(maxSize)))+2; i++ {
		tempLevels = append(tempLevels, i)
	}
	sl.levels = &tempLevels

	tempProbabilities := make([]float64, len(*sl.levels))
	for index := range tempProbabilities {
		if index == 0 {
			tempProbabilities[index] = 1.0
		} else {
			tempProbabilities[index] = 0.5 * tempProbabilities[index-1]
		}
	}
	sl.probabilities = &tempProbabilities

	sl.sample = &Sample{}
	sl.sample.init(sl.levels, sl.probabilities)

	node := SLNode{}

	defaultKey := "__head__"
	defaultLen := len(*sl.levels)
	node.init(&defaultKey, nil, &defaultLen)

	for _, elem := range *sl.levels {
		node.nextPointers[elem] = nil
		node.prevPointers[elem] = nil
	}

	sl.head, sl.tail = &node, &node

	sl.nodeMap = make(map[string]*SLNode)
}
func (sl *SkipList) SearchKey(key string) (int, bool) {
	if elem, ok := sl.nodeMap[key]; ok {
		return *(elem.val), true
	}
	return 0, false
}
func (sl *SkipList) SearchVal(val int) bool {
	currentLevel := (*sl.levels)[len(*sl.levels)-1]
	node := sl.head

	for i := currentLevel; i > 0; i-- {
		for node.nextPointers[i] != nil && *(node.nextPointers[i].val) < val {
			node = node.nextPointers[i]
		}
	}
	return node.nextPointers[1] != nil && *(node.nextPointers[1].val) == val
}
func (sl *SkipList) Insert(key string, val int) {
	if _, ok := sl.nodeMap[key]; !ok {
		currentLevel := (*sl.levels)[len(*sl.levels)-1]
		node := sl.head

		randomLevel := sl.sample.getOne()
		updates := make([]*SLNode, randomLevel)
		for index := range updates {
			updates[index] = nil
		}

		for i := currentLevel; i > 0; i-- {
			for node.nextPointers[i] != nil && *(node.nextPointers[i].val) < val {
				node = node.nextPointers[i]
			}
			if i <= randomLevel {
				updates[i-1] = node
			}
		}
		newNode := SLNode{}
		newNode.init(&key, &val, &randomLevel)

		for i := 1; i <= randomLevel; i++ {
			node = updates[i-1]

			newNode.nextPointers[i] = node.nextPointers[i]

			if node.nextPointers[i] != nil {
				node.nextPointers[i].prevPointers[i] = &newNode
			} else {
				if i == 1 {
					sl.tail = &newNode
				}
			}
			node.nextPointers[i] = &newNode
			newNode.prevPointers[i] = node
		}
		sl.nodeMap[key] = &newNode
	} else {
		sl.Delete(key)
		sl.Insert(key, val)
	}
}
func (sl *SkipList) Delete(key string) bool {
	if elem, ok := sl.nodeMap[key]; ok {
		node := elem

		for i := 1; i <= *(node.level); i++ {
			if node.prevPointers[i] != nil {
				node.prevPointers[i].nextPointers[i] = node.nextPointers[i]
			}
			if node.nextPointers[i] != nil {
				node.nextPointers[i].prevPointers[i] = node.prevPointers[i]
			} else {
				if i == 1 {
					sl.tail = node.prevPointers[i]
				}
			}
		}
		delete(sl.nodeMap, key)
		return true
	}
	return false
}

func (sl *SkipList) GetMax() (string, int, bool) {
	if sl.tail != nil {
		if retKey, retVal := sl.tail.key, sl.tail.val; retKey != nil && retVal != nil {
			return *retKey, *retVal, true
		}
	}
	return "", 1, false
}
func (sl *SkipList) PopMax() (string, int, bool) {
	maxElemKey, maxElemVal, maxElemStatus := sl.GetMax()
	if maxElemStatus != false {
		sl.Delete(maxElemKey)
	}
	return maxElemKey, maxElemVal, maxElemStatus
}

func (sl *SkipList) GetMin() (string, int, bool) {
	if sl.head != nil && sl.head.nextPointers[1] != nil {
		if retKey, retVal := sl.head.nextPointers[1].key, sl.head.nextPointers[1].val; retKey != nil && retVal != nil {
			return *retKey, *retVal, true
		}
	}
	return "", 1, false
}
func (sl *SkipList) PopMin() (string, int, bool) {
	maxElemKey, maxElemVal, maxElemStatus := sl.GetMin()
	if maxElemStatus != false {
		sl.Delete(maxElemKey)
	}
	return maxElemKey, maxElemVal, maxElemStatus
}

func (sl *SkipList) GetPredecessor(val int) (string, int, bool) {
	currentLevel := (*sl.levels)[len(*sl.levels)-1]
	node := sl.head

	for i := currentLevel; i > 0; i-- {
		for node.nextPointers[i] != nil && *(node.nextPointers[i].val) < val {
			node = node.nextPointers[i]
		}
	}
	if node != nil {
		if retKey, retVal := node.key, node.val; retKey != nil && retVal != nil {
			return *retKey, *retVal, true
		}
	}
	return "", 1, false
}
func (sl *SkipList) GetSuccessor(val int) (string, int, bool) {
	currentLevel := (*sl.levels)[len(*sl.levels)-1]
	node := sl.head

	for i := currentLevel; i > 0; i-- {
		for node.nextPointers[i] != nil && *(node.nextPointers[i].val) < val {
			node = node.nextPointers[i]
		}
	}

	if node.nextPointers[1] != nil {
		if *(node.nextPointers[1].val) > val {
			if retKey, retVal := node.nextPointers[1].key, node.nextPointers[1].val; retKey != nil && retVal != nil {
				return *retKey, *retVal, true
			}
		} else {
			if node.nextPointers[1].nextPointers[1] != nil {
				if retKey, retVal := node.nextPointers[1].nextPointers[1].key, node.nextPointers[1].nextPointers[1].val; retKey != nil && retVal != nil {
					return *retKey, *retVal, true
				}
			}
		}
	}
	return "", 1, false
}
func (sl *SkipList) PrintList() {
	for i := len(*sl.levels) - 1; i >= 0; i-- {
		p := sl.head
		out := make([]string, 0)

		for p != nil {
			k, v := "", 0

			if pkey, pval := p.key, p.val; pkey != nil || pval != nil {
				if pkey != nil {
					k = *pkey
				}
				if pval != nil {
					v = *pval
				}
			}
			out = append(out, fmt.Sprintf("%v <-> %v", k, v))
			p = p.nextPointers[(*sl.levels)[i]]
		}

		//for _, elem := range out {
		fmt.Println(out)
		//fmt.Println()
		//}
	}
}

func runner() {
	sl := SkipList{}
	sl.init(1000000)

	sl.Insert("hello", 3)
	sl.Insert("world", 11)
	sl.Insert("this", 18)
	sl.Insert("is", 9)
	sl.Insert("me", 14)
	sl.Insert("reaching", 19)
	sl.Insert("out", 17)
	sl.Insert("to", 6)
	sl.Insert("you", 24)
	sl.Insert("to", 27)
	sl.Insert("talk", 10)
	sl.Insert("about", 2)
	sl.Insert("the", 4)
	sl.Insert("extended", 26)
	sl.Insert("warranty", 29)
	sl.Insert("of", 25)
	sl.Insert("your", 23)
	sl.Insert("car", 21)
	sl.Insert("insaurance", 22)
	sl.Insert("please", 20)
	sl.Insert("response", 5)
	sl.Insert("ASAP", 12)
	sl.Insert("because", 13)
	sl.Insert("this", 1)
	sl.Insert("is", 28)
	sl.Insert("a", 16)
	sl.Insert("limited", 8)
	sl.Insert("time", 15)
	sl.Insert("offer", 7)

	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())

	fmt.Println(sl.SearchVal(20))

	/*	fmt.Println(sl.SearchVal("world"))
		fmt.Println(sl.SearchVal("this"))
		fmt.Println(sl.SearchVal("is"))
		fmt.Println(sl.SearchVal("me"))
		fmt.Println(sl.SearchVal("reaching"))
		fmt.Println(sl.SearchVal("out"))
		fmt.Println(sl.SearchVal("to"))
		fmt.Println(sl.SearchVal("you"))
		fmt.Println(sl.SearchVal("to"))
		fmt.Println(sl.SearchVal("talk"))
		fmt.Println(sl.SearchVal("about"))
		fmt.Println(sl.SearchVal("the"))
		fmt.Println(sl.SearchVal("extended"))
		fmt.Println(sl.SearchVal("warranty"))
		fmt.Println(sl.SearchVal("of"))
		fmt.Println(sl.SearchVal("your"))
		fmt.Println(sl.SearchVal("car"))
		fmt.Println(sl.SearchVal("insaurance"))
		fmt.Println(sl.SearchVal("please"))
		fmt.Println(sl.SearchVal("response"))
		fmt.Println(sl.SearchVal("ASAP"))
		fmt.Println(sl.SearchVal("because"))
		fmt.Println(sl.SearchVal("this"))
		fmt.Println(sl.SearchVal("is"))
		fmt.Println(sl.SearchVal("a"))
		fmt.Println(sl.SearchVal("limited"))
		fmt.Println(sl.SearchVal("time"))
		fmt.Println(sl.SearchVal("offer"))
		fmt.Println(sl.SearchVal("SUS"))
	*/

	//sl.PrintList()

	/*fmt.Println(sl.PopMin())
	fmt.Println(sl.PopMin())
	fmt.Println(sl.PopMin())
	fmt.Println(sl.PopMin())

	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())
	fmt.Println(sl.PopMax())*/

	fmt.Println(sl.GetPredecessor(2))
	fmt.Println(sl.GetPredecessor(13))
	fmt.Println(sl.GetPredecessor(20))

}

File: scripts/skiplist.py
import math
import random


class Sample(object):
    def __init__(self, values: list[int], weights: list[float]):
        self.values = values
        self.weights = weights

        s = sum(self.weights)
        self.weights = [x / float(s) if s != 0 else 0.0 for x in self.weights]

        self.cum_sum = [0] * len(self.values)
        self.compute_cum_sum()

    def compute_cum_sum(self):
        cum_sum = [0] * len(self.values)
        s = 0
        for i in range(len(self.values)):
            s += self.weights[i]
            cum_sum[i] = s

        self.cum_sum = cum_sum

    def get_one(self):
        u = random.uniform(0, 1)

        left, right = 0, len(self.values) - 1
        last_true = 0
        while left <= right:
            mid = int((left + right) / 2)
            if self.cum_sum[mid] >= u:
                last_true = mid
                right = mid - 1
            else:
                left = mid + 1

        return self.values[last_true]


class Node(object):
    def __init__(self, key=None, val=None, level=None):
        self.key = key
        self.val = val
        self.next_pointers = {}
        self.prev_pointers = {}
        self.level = level


class SkipList(object):
    def __init__(self, max_size=10 ** 6):
        self.levels = range(1, int(math.log(max_size, 2)) + 2)

        self.probabilities = [1.0] * len(self.levels)
        for i in range(len(self.levels)):
            self.probabilities[i] = 0.5 * self.probabilities[i - 1] if i > 0 else 1.0

        self.sample = Sample(self.levels, self.probabilities)
        node = Node('__head__', -float("Inf"), len(self.levels))

        for i in self.levels:
            node.next_pointers[i] = None
            node.prev_pointers[i] = None

        self.head = node
        self.tail = node

        self.node_map = {}

    def search_key(self, key):
        return self.node_map[key].val if key in self.node_map else False

    def search_val(self, val):
        curr_level = self.levels[-1]
        node = self.head

        for i in range(curr_level, 0, -1):
            while node.next_pointers[i] is not None and node.next_pointers[i].val < val:
                node = node.next_pointers[i]

        return node.next_pointers[1] is not None and node.next_pointers[1].val == val

    def insert(self, key, val):
        if key not in self.node_map:
            curr_level = self.levels[-1]
            node = self.head

            random_level = self.sample.get_one()
            updates = [None] * random_level

            for i in range(curr_level, 0, -1):
                while node.next_pointers[i] is not None and node.next_pointers[i].val < val:
                    node = node.next_pointers[i]

                if i <= random_level:
                    updates[i - 1] = node

            new_node = Node(key, val, random_level)

            for i in range(1, random_level + 1):
                node = updates[i - 1]

                new_node.next_pointers[i] = node.next_pointers[i]

                if node.next_pointers[i] is not None:
                    node.next_pointers[i].prev_pointers[i] = new_node
                else:
                    if i == 1:
                        self.tail = new_node

                node.next_pointers[i] = new_node
                new_node.prev_pointers[i] = node

            self.node_map[key] = new_node

        else:
            self.delete(key)
            self.insert(key, val)

    def delete(self, key):
        if key in self.node_map:
            node = self.node_map[key]

            for i in range(1, node.level + 1):
                if node.prev_pointers[i] is not None:
                    node.prev_pointers[i].next_pointers[i] = node.next_pointers[i]

                if node.next_pointers[i] is not None:
                    node.next_pointers[i].prev_pointers[i] = node.prev_pointers[i]
                else:
                    if i == 1:
                        self.tail = node.prev_pointers[i]

            self.node_map.pop(key)

    def get_min(self):
        if self.head is not None and self.head.next_pointers[1] is not None:
            return self.head.next_pointers[1].key, self.head.next_pointers[1].val
        return None

    def get_max(self):
        if self.tail is not None:
            return self.tail.key, self.tail.val
        return None

    def pop_min(self):
        out = self.get_min()
        if out is not None:
            self.delete(out[0])
        return out

    def pop_max(self):
        out = self.get_max()
        if out is not None:
            self.delete(out[0])
        return out

    def get_predecessor(self, val):
        curr_level = self.levels[-1]
        node = self.head

        for i in range(curr_level, 0, -1):
            while node.next_pointers[i] is not None and node.next_pointers[i].val < val:
                node = node.next_pointers[i]

        if node is not None:
            return node.key, node.val

        return None

    def get_successor(self, val):
        curr_level = self.levels[-1]
        node = self.head

        for i in range(curr_level, 0, -1):
            while node.next_pointers[i] is not None and node.next_pointers[i].val < val:
                node = node.next_pointers[i]

        if node.next_pointers[1] is not None:
            if node.next_pointers[1].val > val:
                return node.next_pointers[1].key, node.next_pointers[1].val
            else:
                if node.next_pointers[1].next_pointers[1] is not None:
                    return node.next_pointers[1].next_pointers[1].key, node.next_pointers[1].next_pointers[1].val
        return None

    def print_list(self):
        for i in range(len(self.levels) - 1, -1, -1):
            p = self.head
            out = []

            while p is not None:
                out.append((p.key, p.val))
                p = p.next_pointers[self.levels[i]]

            print(out)


sl = SkipList()

sl.insert(3, 'hello')
sl.insert(11, 'world')
sl.insert(18, 'this')
sl.insert(9, 'is')
sl.insert(14, 'me')
sl.insert(19, 'reaching')
sl.insert(17, 'out')
sl.insert(6, 'to')
sl.insert(24, 'you')
sl.insert(27, 'to')
sl.insert(10, 'talk')
sl.insert(2, 'about')
sl.insert(4, 'the')
sl.insert(26, 'extended')
sl.insert(29, 'warranty')
sl.insert(25, 'of')
sl.insert(23, 'your')
sl.insert(21, 'car')
sl.insert(22, 'insaurance')
sl.insert(20, 'please')
sl.insert(5, 'response')
sl.insert(12, 'ASAP')
sl.insert(13, 'because')
sl.insert(1, 'this')
sl.insert(28, 'is')
sl.insert(16, 'a')
sl.insert(8, 'limited')
sl.insert(15, 'time')
sl.insert(7, 'offer')

print(sl.search_val('hello'))
# sl.delete
print(sl.search_val('world'))
print(sl.search_val('this'))
print(sl.search_val('is'))
print(sl.search_val('me'))
print(sl.search_val('reaching'))
print(sl.search_val('out'))
print(sl.search_val('to'))
print(sl.search_val('you'))
print(sl.search_val('to'))
print(sl.search_val('talk'))
print(sl.search_val('about'))
print(sl.search_val('the'))
print(sl.search_val('extended'))
print(sl.search_val('warranty'))
print(sl.search_val('of'))
print(sl.search_val('your'))
print(sl.search_val('car'))
print(sl.search_val('insaurance'))
print(sl.search_val('please'))
print(sl.search_val('response'))
print(sl.search_val('ASAP'))
print(sl.search_val('because'))
print(sl.search_val('this'))
print(sl.search_val('is'))
print(sl.search_val('a'))
print(sl.search_val('limited'))
print(sl.search_val('time'))
print(sl.search_val('offer'))

sl.print_list()
